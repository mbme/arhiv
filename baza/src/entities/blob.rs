use anyhow::Result;

use rs_utils::{file_exists, get_file_size, get_media_type};

use crate::entities::BLOBId;

#[derive(Hash, Eq, PartialEq)]
pub struct BLOB {
    pub id: BLOBId,
    pub file_path: String,
}

impl BLOB {
    #[must_use]
    pub fn new(id: BLOBId, data_dir: &str) -> Self {
        BLOB {
            file_path: format!("{}/{}", data_dir, id.get_file_name()),
            id,
        }
    }

    pub fn exists(&self) -> Result<bool> {
        file_exists(&self.file_path)
    }

    pub fn get_size(&self) -> Result<u64> {
        get_file_size(&self.file_path)
    }

    pub fn get_media_type(&self) -> Result<String> {
        get_media_type(&self.file_path)
    }
}
