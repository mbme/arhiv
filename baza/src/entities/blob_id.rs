use std::{fmt, ops::Deref};

use anyhow::{ensure, Context, Result};
use serde::{Deserialize, Serialize};

use rs_utils::{get_file_hash_blake3, is_valid_base64, to_url_safe_base64};

const BLOB_FILE_PREFIX: &str = "blake3-";
const BLOB_ID_EXPECTED_LENGTH: usize = 44;

#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, Hash)]
pub struct BLOBId(String);

impl fmt::Display for BLOBId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl Deref for BLOBId {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl BLOBId {
    pub fn is_valid_blob_id(id: &str) -> Result<()> {
        ensure!(
            id.len() == BLOB_ID_EXPECTED_LENGTH,
            "BLOB id must be {} chars long",
            BLOB_ID_EXPECTED_LENGTH
        );

        ensure!(is_valid_base64(id), "BLOB id must be a valid base64 string");

        Ok(())
    }

    pub fn from_file(file_path: &str) -> Result<BLOBId> {
        let hash = get_file_hash_blake3(file_path).context("failed to calculate file hash")?;

        let id = to_url_safe_base64(&hash);

        Ok(BLOBId(id))
    }

    #[must_use]
    pub fn get_file_name(&self) -> String {
        format!("{}{}", BLOB_FILE_PREFIX, self.0)
    }

    pub fn from_file_name(file_name: &str) -> Result<BLOBId> {
        ensure!(
            file_name.starts_with(BLOB_FILE_PREFIX),
            "BLOB file name must start with the prefix {}",
            BLOB_FILE_PREFIX
        );
        let blob_id = &file_name[BLOB_FILE_PREFIX.len()..];

        BLOBId::is_valid_blob_id(blob_id)?;

        let blob_id = BLOBId::from_string(blob_id);

        Ok(blob_id)
    }

    #[must_use]
    pub fn from_string(blob_id: impl Into<String>) -> BLOBId {
        BLOBId(blob_id.into())
    }
}
