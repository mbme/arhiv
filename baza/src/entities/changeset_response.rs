use std::fmt;

use anyhow::{anyhow, Context, Result};
use serde::{Deserialize, Serialize};

use crate::sync::revision::Revision;

use super::Document;

#[derive(Serialize, Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct ChangesetResponse {
    // replica revision
    pub base_rev: Revision,

    // primary revision
    pub latest_rev: Revision,

    // documents_history records with rev > replica_rev
    pub new_snapshots: Vec<Document>,

    pub conflicts: Vec<Document>,
}

impl ChangesetResponse {
    #[must_use]
    pub fn serialize(&self) -> String {
        serde_json::to_string(self).expect("Failed to serialize ChangesetResponse to json")
    }
}

impl std::str::FromStr for ChangesetResponse {
    type Err = anyhow::Error;

    fn from_str(data: &str) -> Result<ChangesetResponse> {
        serde_json::from_str(data).context(anyhow!("Failed to parse ChangesetResponse:\n{}", data))
    }
}

impl fmt::Display for ChangesetResponse {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "[ChangesetResponse: rev {}, {} snapshot(s) and {} conflict(s) since {}]",
            self.latest_rev,
            self.new_snapshots.len(),
            self.conflicts.len(),
            self.base_rev,
        )
    }
}
