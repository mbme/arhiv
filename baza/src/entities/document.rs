use std::fmt;

use anyhow::{Context, Result};
use serde::{de::DeserializeOwned, Deserialize, Serialize};

use rs_utils::{now, Timestamp};

use super::{DocumentClass, DocumentData, Id, Revision};

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct Document<D = DocumentData> {
    pub id: Id,
    pub rev: Option<Revision>,
    #[serde(flatten)]
    pub class: DocumentClass,
    pub updated_at: Timestamp,
    pub data: D,
}

impl<D> Document<D> {
    #[must_use]
    pub fn new_with_data(class: DocumentClass, data: D) -> Self {
        Document {
            id: Id::new(),
            rev: None,
            class,
            updated_at: now(),
            data,
        }
    }
}

impl Document {
    #[must_use]
    pub fn new(class: DocumentClass) -> Self {
        Document::new_with_data(class, DocumentData::new())
    }

    #[must_use]
    pub fn is_erased(&self) -> bool {
        self.class.is_erased()
    }

    #[must_use]
    pub fn is_staged(&self) -> bool {
        self.rev.is_none()
    }

    #[must_use]
    pub fn is_committed(&self) -> bool {
        !self.is_staged()
    }

    pub fn get_rev(&self) -> Result<&Revision> {
        self.rev.as_ref().context("document revision is missing")
    }

    pub(crate) fn erase(&mut self) {
        self.class = DocumentClass::erased();
        self.data = DocumentData::new();
        self.updated_at = now();
    }

    pub(crate) fn stage(&mut self) {
        self.rev = None;
    }

    pub fn convert<D: DeserializeOwned>(self) -> Result<Document<D>> {
        let data: D = serde_json::from_value(self.data.into()).context("failed to convert data")?;

        Ok(Document {
            id: self.id,
            rev: self.rev,
            class: self.class,
            updated_at: self.updated_at,
            data,
        })
    }
}

impl<D: Serialize> Document<D> {
    pub fn into_document(self) -> Result<Document> {
        let data: DocumentData = serde_json::to_value(self.data)
            .context("failed to convert to value")?
            .try_into()?;

        Ok(Document {
            id: self.id,
            rev: self.rev,
            class: self.class,
            updated_at: self.updated_at,
            data,
        })
    }
}

impl std::str::FromStr for Document {
    type Err = anyhow::Error;

    fn from_str(data: &str) -> Result<Document> {
        serde_json::from_str(data).context("Failed to parse document json")
    }
}

impl fmt::Display for Document {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "[Document {} {} {}]",
            self.class,
            self.id,
            Revision::to_string(&self.rev),
        )
    }
}
