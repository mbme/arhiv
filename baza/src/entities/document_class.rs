use std::fmt;

use serde::{Deserialize, Serialize};

pub const ERASED_DOCUMENT_TYPE: &str = "";

#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, Hash)]
pub struct DocumentClass {
    pub document_type: String,
    pub subtype: String,
}

impl DocumentClass {
    pub fn erased() -> Self {
        DocumentClass {
            document_type: ERASED_DOCUMENT_TYPE.to_string(),
            subtype: "".to_string(),
        }
    }

    pub fn new(document_type: impl Into<String>, subtype: impl Into<String>) -> Self {
        DocumentClass {
            document_type: document_type.into(),
            subtype: subtype.into(),
        }
    }

    pub fn set_subtype(&mut self, subtype: impl Into<String>) {
        self.subtype = subtype.into();
    }

    #[must_use]
    pub fn is_erased(&self) -> bool {
        self.document_type == ERASED_DOCUMENT_TYPE
    }
}

impl fmt::Display for DocumentClass {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match (self.document_type.as_str(), self.subtype.as_str()) {
            (ERASED_DOCUMENT_TYPE, _) => write!(f, "erased"),
            (document_type, "") => write!(f, "{document_type}"),
            (document_type, subtype) => write!(f, "{document_type}/{subtype}"),
        }
    }
}
