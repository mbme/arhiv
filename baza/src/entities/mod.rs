mod document;
mod document_data;
mod document_key;
mod document_lock;
mod document_type;
mod id;
mod instance_id;
mod refs;
mod revision;

pub use document::Document;
pub use document_data::DocumentData;
pub use document_key::DocumentKey;
pub use document_lock::{DocumentLock, DocumentLockKey};
pub use document_type::{DocumentType, ERASED_DOCUMENT_TYPE};
pub use id::Id;
pub use instance_id::InstanceId;
pub use refs::Refs;
pub use revision::{LatestRevComputer, Revision, VectorClockOrder};

#[cfg(test)]
pub use document::{new_document, new_empty_document};
