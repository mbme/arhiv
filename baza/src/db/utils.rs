use anyhow::{anyhow, Context, Result};
use rusqlite::Row;
use serde_json::Value;

use crate::entities::{BLOBId, Document, DocumentClass};

pub fn extract_document(row: &Row) -> Result<Document> {
    let document_type: String = row.get("document_type")?;
    let subtype: String = row.get("subtype")?;

    Ok(Document {
        id: row.get("id")?,
        rev: {
            let rev: Value = row.get("rev")?;

            rev.try_into()?
        },
        class: DocumentClass::new(document_type, subtype),
        created_at: row.get("created_at")?,
        updated_at: row.get("updated_at")?,
        data: {
            let data: Value = row.get("data")?;

            data.try_into()?
        },
    })
}

pub fn extract_blob_id(row: &Row) -> Result<BLOBId> {
    row.get("blob_id")
        .context(anyhow!("failed to extract blob_id"))
}
