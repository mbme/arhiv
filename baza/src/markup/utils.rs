use crate::entities::Id;

const REF_LINK_PREFIX: &str = "ref:";

#[must_use]
pub fn extract_id(text: &str) -> Option<Id> {
    if text.starts_with(REF_LINK_PREFIX) && text.len() > REF_LINK_PREFIX.len() {
        let id = text[REF_LINK_PREFIX.len()..].into();

        Some(id)
    } else {
        None
    }
}

#[must_use]
pub fn create_link(url: &str, text: &str) -> String {
    if text.is_empty() {
        format!("<{url}>")
    } else {
        format!("[{text}]({url})")
    }
}

#[must_use]
pub fn create_ref(id: &Id, text: &str) -> String {
    create_link(&format!("{REF_LINK_PREFIX}{id}"), text)
}
