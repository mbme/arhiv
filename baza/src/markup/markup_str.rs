use std::convert::From;
use std::{collections::HashSet, ops::Range};

use anyhow::{bail, ensure, Context, Result};
use pulldown_cmark::{
    Alignment, CodeBlockKind, CowStr, Event, HeadingLevel, LinkType, Options, Parser, Tag,
};
use serde::Serialize;

use rs_utils::create_byte_pos_to_char_pos_map;

use super::utils::extract_id;
use crate::entities::*;

pub struct MarkupStr<'a>(CowStr<'a>);

impl MarkupStr<'_> {
    #[must_use]
    pub fn parse(&self) -> Parser {
        let mut options = Options::empty();
        options.insert(Options::ENABLE_STRIKETHROUGH);
        options.insert(Options::ENABLE_TABLES);
        options.insert(Options::ENABLE_TASKLISTS);

        Parser::new_ext(self.0.as_ref(), options)
    }

    #[must_use]
    pub fn extract_refs(&self) -> HashSet<Id> {
        let mut refs: HashSet<Id> = HashSet::new();

        let parser = self.parse();
        for event in parser {
            if let Event::Start(Tag::Link { ref dest_url, .. })
            | Event::Start(Tag::Image { ref dest_url, .. }) = event
            {
                if let Some(id) = extract_id(dest_url) {
                    refs.insert(id);
                }
            }
        }

        refs
    }

    #[must_use]
    pub fn preview(&self, lines: usize) -> Self {
        self.0
            .as_ref()
            .lines()
            .take(lines)
            .collect::<Vec<_>>()
            .join("\n")
            .into()
    }

    pub fn get_ast(&self) -> Result<MarkupElement<'_>> {
        let mut stack: Vec<MarkupElement<'_>> = vec![MarkupElement::Document { children: vec![] }];

        let byte_pos_to_char_pos_map = create_byte_pos_to_char_pos_map(&self.0);

        for (event, byte_range) in self.parse().into_offset_iter() {
            let range = Range {
                start: *byte_pos_to_char_pos_map
                    .get(&byte_range.start)
                    .context(format!(
                        "failed to convert range start {} to char",
                        byte_range.start
                    ))?,
                end: *byte_pos_to_char_pos_map
                    .get(&byte_range.end)
                    .context(format!(
                        "failed to convert range end {} to char",
                        byte_range.end
                    ))?,
            };

            match event {
                Event::Text(value) => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::Text { value, range })?;
                }
                Event::Code(value) => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::Code { value, range })?;
                }
                Event::Html(value) => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::Html { value, range })?;
                }
                Event::FootnoteReference(label) => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::FootnoteReference { label, range })?;
                }
                Event::SoftBreak => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::SoftBreak { range })?;
                }
                Event::HardBreak => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::HardBreak { range })?;
                }
                Event::Rule => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::Rule { range })?;
                }
                Event::TaskListMarker(checked) => {
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(MarkupElement::TaskListMarker { checked, range })?;
                }

                Event::Start(tag) => {
                    let el = match tag {
                        Tag::Paragraph => MarkupElement::Paragraph {
                            children: vec![],
                            range,
                        },
                        Tag::Heading { level, .. } => MarkupElement::Heading {
                            range,
                            level,
                            children: vec![],
                        },
                        Tag::BlockQuote(_) => MarkupElement::BlockQuote {
                            children: vec![],
                            range,
                        },
                        Tag::CodeBlock(kind) => MarkupElement::CodeBlock {
                            range,
                            kind,
                            children: vec![],
                        },
                        Tag::List(first_item_number) => MarkupElement::List {
                            range,
                            first_item_number,
                            children: vec![],
                        },
                        Tag::Item => MarkupElement::ListItem {
                            children: vec![],
                            range,
                        },
                        Tag::FootnoteDefinition(label) => MarkupElement::FootnoteDefinition {
                            range,
                            label,
                            children: vec![],
                        },
                        Tag::Table(alignments) => MarkupElement::Table {
                            range,
                            alignments,
                            children: vec![],
                        },
                        Tag::TableHead => MarkupElement::TableHead {
                            children: vec![],
                            range,
                        },
                        Tag::TableRow => MarkupElement::TableRow {
                            children: vec![],
                            range,
                        },
                        Tag::TableCell => MarkupElement::TableCell {
                            children: vec![],
                            range,
                        },

                        Tag::Emphasis => MarkupElement::Emphasis {
                            children: vec![],
                            range,
                        },
                        Tag::Strong => MarkupElement::Strong {
                            children: vec![],
                            range,
                        },
                        Tag::Strikethrough => MarkupElement::Strikethrough {
                            children: vec![],
                            range,
                        },

                        Tag::Link {
                            link_type,
                            dest_url,
                            ..
                        } => MarkupElement::Link {
                            link_type,
                            url: dest_url,
                            children: vec![],
                            range,
                        },

                        Tag::Image {
                            link_type,
                            dest_url,
                            ..
                        } => MarkupElement::Image {
                            link_type,
                            url: dest_url,
                            children: vec![],
                            range,
                        },
                        Tag::HtmlBlock => break,
                        Tag::MetadataBlock(_) => break,
                        Tag::DefinitionList => break,
                        Tag::DefinitionListTitle => break,
                        Tag::DefinitionListDefinition => break,
                        Tag::Superscript => break,
                        Tag::Subscript => break,
                    };

                    stack.push(el);
                }

                Event::End(_tag) => {
                    let el = stack.pop().context("stack must not be empty")?;
                    stack
                        .last_mut()
                        .context("stack must not be empty")?
                        .add_child(el)?;
                }
                Event::InlineHtml(_) => {}
                Event::InlineMath(_) => {}
                Event::DisplayMath(_) => {}
            }
        }

        ensure!(stack.len() == 1, "only Document must be on stack");

        stack.pop().context("Document must be on stack")
    }
}

impl<'a> From<&'a str> for MarkupStr<'a> {
    fn from(value: &'a str) -> Self {
        MarkupStr(CowStr::Borrowed(value))
    }
}

impl From<String> for MarkupStr<'_> {
    fn from(value: String) -> Self {
        MarkupStr(CowStr::Boxed(value.into_boxed_str()))
    }
}

pub type Children<'a> = Vec<MarkupElement<'a>>;

#[derive(Serialize, Debug)]
#[serde(deny_unknown_fields, tag = "typeName")]
pub enum MarkupElement<'a> {
    Document {
        children: Children<'a>,
    },
    Text {
        range: Range<usize>,
        value: CowStr<'a>,
    },
    Code {
        range: Range<usize>,
        value: CowStr<'a>,
    },
    Html {
        range: Range<usize>,
        value: CowStr<'a>,
    },
    FootnoteReference {
        range: Range<usize>,
        label: CowStr<'a>,
    },
    SoftBreak {
        range: Range<usize>,
    },
    HardBreak {
        range: Range<usize>,
    },
    Rule {
        range: Range<usize>,
    },
    TaskListMarker {
        range: Range<usize>,
        checked: bool,
    },
    Paragraph {
        range: Range<usize>,
        children: Children<'a>,
    },
    Heading {
        range: Range<usize>,
        level: HeadingLevel,
        children: Children<'a>,
    },
    BlockQuote {
        range: Range<usize>,
        children: Children<'a>,
    },
    CodeBlock {
        range: Range<usize>,
        kind: CodeBlockKind<'a>,
        children: Children<'a>,
    },
    List {
        range: Range<usize>,
        first_item_number: Option<u64>,
        children: Children<'a>,
    },
    ListItem {
        range: Range<usize>,
        children: Children<'a>,
    },
    FootnoteDefinition {
        range: Range<usize>,
        label: CowStr<'a>,
        children: Children<'a>,
    },
    Table {
        range: Range<usize>,
        alignments: Vec<Alignment>,
        children: Children<'a>,
    },
    TableHead {
        range: Range<usize>,
        children: Children<'a>, // only TableCell
    },
    TableRow {
        range: Range<usize>,
        children: Children<'a>, // only TableCell
    },
    TableCell {
        range: Range<usize>,
        children: Children<'a>,
    },

    Emphasis {
        range: Range<usize>,
        children: Children<'a>,
    },
    Strong {
        range: Range<usize>,
        children: Children<'a>,
    },
    Strikethrough {
        range: Range<usize>,
        children: Children<'a>,
    },

    Link {
        range: Range<usize>,
        link_type: LinkType,
        url: CowStr<'a>,
        children: Children<'a>,
    },

    Image {
        range: Range<usize>,
        link_type: LinkType,
        url: CowStr<'a>,
        children: Children<'a>,
    },
}

impl<'a> MarkupElement<'a> {
    fn get_children_mut(&mut self) -> Option<&mut Children<'a>> {
        match self {
            MarkupElement::Document { children }
            | MarkupElement::Paragraph { children, .. }
            | MarkupElement::Heading { children, .. }
            | MarkupElement::BlockQuote { children, .. }
            | MarkupElement::CodeBlock { children, .. }
            | MarkupElement::List { children, .. }
            | MarkupElement::ListItem { children, .. }
            | MarkupElement::FootnoteDefinition { children, .. }
            | MarkupElement::Table { children, .. }
            | MarkupElement::TableHead { children, .. }
            | MarkupElement::TableRow { children, .. }
            | MarkupElement::TableCell { children, .. }
            | MarkupElement::Emphasis { children, .. }
            | MarkupElement::Strong { children, .. }
            | MarkupElement::Strikethrough { children, .. }
            | MarkupElement::Link { children, .. }
            | MarkupElement::Image { children, .. } => {
                //
                Some(children)
            }

            _ => None,
        }
    }

    fn add_child(&mut self, child: MarkupElement<'a>) -> Result<()> {
        if let Some(children) = self.get_children_mut() {
            children.push(child);

            Ok(())
        } else {
            bail!("can't push child into {:?}", self)
        }
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    use serde_json::Value;

    use crate::{
        entities::Id,
        markup::{create_image_ref, create_ref, MarkupStr},
    };

    fn into_ast(value: &str) -> Value {
        let markup = MarkupStr::from(value);

        let ast = markup.get_ast().expect("must get ast");

        serde_json::to_value(ast).expect("ast must serialize")
    }

    #[test]
    fn test_parse_header() {
        let ast = into_ast("# TEST HEADER");

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_parse_list() {
        let ast = into_ast(
            "
- [ ] test 1
- [+] test 2
- [x] test 3
- test 4
* test 5
",
        );

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_parse_hr() {
        let ast = into_ast("---");

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_parse_table() {
        let ast = into_ast(
            "
| Syntax      | Description | Test Text     |
| :---        |    :----:   |          ---: |
| Header      | Title       | Here's this   |
| Paragraph   | Text        | And more      |
",
        );

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_parse_code_block() {
        let ast = into_ast(
            r#"
```json
{
  "firstName": "John",
  "lastName": "Smith",
  "age": 25
}
```
"#,
        );

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_parse_link() {
        let ast = into_ast(r#"[test](http://example.com "some title")"#);

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_parse_autolink() {
        let ast = into_ast(r#"<http://example.com>"#);

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_parse_image() {
        let ast = into_ast(r#"![test](http://example.com "some title")"#);

        insta::assert_json_snapshot!(ast);
    }

    #[test]
    fn test_extract_refs() {
        let id1 = Id::new();
        let id2 = Id::new();
        let id3 = Id::new();

        let ref1 = create_ref(&id1, "");
        let ref2 = create_ref(&id2, "test");
        let ref3 = create_image_ref(&id3, "test");

        let markup = MarkupStr::from(format!("{ref1}    {ref2}     {ref3}"));

        let refs = markup.extract_refs();

        assert_eq!(refs, HashSet::from_iter([id1, id2, id3]));
    }
}
