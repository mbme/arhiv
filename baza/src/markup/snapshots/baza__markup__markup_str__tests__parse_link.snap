---
source: baza/src/markup/markup_str.rs
assertion_line: 450
expression: ast
---
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "range": {
                "end": 5,
                "start": 1
              },
              "typeName": "Text",
              "value": "test"
            }
          ],
          "link_type": "Inline",
          "range": {
            "end": 39,
            "start": 0
          },
          "typeName": "Link",
          "url": "http://example.com"
        }
      ],
      "range": {
        "end": 39,
        "start": 0
      },
      "typeName": "Paragraph"
    }
  ],
  "typeName": "Document"
}
